### PR TITLE
Python examples to construct a SIR

### PR DESCRIPTION
**Pull request by cosunae**
* Reorganize the proto files so that the python API is generated within corresponding SIR/IIR modules. 
* provide a set of python examples to generate HIR code

### CODE REVIEW

**Comment by cosunae** [comment]
launch jenkins

**Comment by cosunae** [comment]
launch jenkins

File: cmake/modules/DawnProtobufGenerate.cmake [hunk 1]
```diff
@@ -129,6 +142,14 @@ function(dawn_protobuf_generate)
     endforeach()
   endforeach()
 
+  if(ARG_OUT_DIRS) 
+    if(${ARG_LANGUAGE} STREQUAL "java" )
```
Comment (twicki) [inline cmake/modules/DawnProtobufGenerate.cmake]: why is there java-stuff in here?

File: examples/python/config.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/python3
```
Comment (twicki) [inline examples/python/config.py]: is this file needed? this should be the generated one?

File: examples/python/copy_stencil.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
+"""Copy stencil HIR generator
```
Comment (twicki) [inline examples/python/copy_stencil.py]: I would generally advise to follow https://www.python.org/dev/peps/pep-0008/ as a python style guide unless you have another preference

File: src/dawn/Compiler/Options.inc [hunk 4]
```diff
@@ -60,7 +60,7 @@ OPT(std::string, ReorderStrategy, "greedy", "reorder", "",
     "\n - scut   = Use S-cut graph partitioning\n", "<strategy>", true, false)
 OPT(int, MaxFieldsPerStencil, 40, "max-fields", "",
     "Set the maximum number of fields in any given stencils", "<N>", true, false)
-OPT(bool, PassVerbose, false, "pass-verbose", "",
+OPT(bool, PassVerbose, true, "pass-verbose", "",
```
Comment (twicki) [inline src/dawn/Compiler/Options.inc]: I wouldn't do that by default

File: src/dawn/Serialization/SIRSerializer.cpp [hunk 5]
```diff
@@ -15,7 +15,7 @@
 #include "dawn/SIR/SIR.h"
 #include "dawn/SIR/AST.h"
 #include "dawn/SIR/ASTVisitor.h"
-#include "dawn/SIR/SIR.pb.h"
+#include "dawn/SIR/SIR/SIR.pb.h"
```
Comment (twicki) [inline src/dawn/Serialization/SIRSerializer.cpp]: What was the intention behind this naming structure? wouldn't `dawn/SIR/proto/SIR.pb.h` be the better path? `SIR/SIR` seems a little redundant

Comment (cosunae) [reply src/dawn/Serialization/SIRSerializer.cpp]: yes, it is redundant. However this is impose by proto buffer, if we want to create a module out of SIR (with SIR and statements) instead of standalone files. I think it is described here
https://github.com/protocolbuffers/protobuf/issues/1491

File: examples/python/config.py.in [hunk 6]
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/python3
+# -*- coding: utf-8 -*-
+##===-----------------------------------------------------------------------------*- Python -*-===##
```
Comment (twicki) [inline examples/python/config.py.in]: python-bock comments are usually just one `#` and not two (https://www.python.org/dev/peps/pep-0008/#block-comments)

File: examples/python/hori_diff.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+internal IR. 
+The program contains two parts: 
+    1. construct the HIR of the example
+    2. pass the HIR to the dawn compiler in order to run all optimizer passes and code generation.
+       In this example the compiler is configured with the CUDA backend, therefore will code generate
+       an optimized CUDA implementation.
+
+"""
+
+import textwrap
```
Comment (twicki) [inline examples/python/hori_diff.py]: same for coding standard here

File: examples/python/copy_stencil.py (hunk 3 above)
Comment (twicki) [reply examples/python/copy_stencil.py]: mostly https://www.python.org/dev/peps/pep-0008/#naming-conventions:

Classes are CamelCase, functions are lowercase with underscores , variables are also lowercase with underscores

File: examples/python/tridiagonal_solve.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
+"""Tridiagonal solve computation HIR generator
```
Comment (twicki) [inline examples/python/tridiagonal_solve.py]: when I try to run this I don't get the correct path out of config. Can you maybe show me either hands on or elaborate what might be wrong?

File: python/IIR/__init__.py [hunk 9]
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/python3
+# -*- coding: utf-8 -*-
+##===-----------------------------------------------------------------------------*- Python -*-===##
```
Comment (twicki) [inline python/IIR/__init__.py]: comment block with only `# `

File: python/dawn/sir_printer.py [hunk 10]
```diff
@@ -0,0 +1,229 @@
+import textwrap
+
+class SIRPrinter:
+    def __init__(self):
+        self.indent_ = 0
+        self.T_ = textwrap.TextWrapper(initial_indent=' ' * self.indent_, width=120,
+                                       subsequent_indent=' ' * self.indent_)
+
+    def visitBuiltinType(self, builtin_type):
```
Comment (twicki) [inline python/dawn/sir_printer.py]: PEP 8

File: src/dawn/Serialization/SIRSerializer.cpp (hunk 5 above)
Comment (twicki) [reply src/dawn/Serialization/SIRSerializer.cpp]: I see, got it

**Comment by cosunae** [comment]
launch jenkins

File: CMakeLists.txt [hunk 11]
```diff
@@ -133,9 +133,16 @@ add_subdirectory(src)
 add_subdirectory(cmake)
 
 if(DAWN_PYTHON)
+  if(NOT DEFINED PYTHON_EXECUTABLE)
```
[the remark this answers is not in the document]
Comment (twicki) [reply CMakeLists.txt]: you're right, sorry

File: examples/python/config.py.in (hunk 6 above)
Comment (cosunae) [reply examples/python/config.py.in]: I moved everything (python) to PEP8 except the makers, since they would break API and rupert is using it at the moment. 
I created another PR for doing this as well

File: examples/python/copy_stencil.py (hunk 3 above)
Comment (cosunae) [reply examples/python/copy_stencil.py]: done

File: examples/python/hori_diff.py (hunk 7 above)
Comment (cosunae) [reply examples/python/hori_diff.py]: done

File: examples/python/tridiagonal_solve.py (hunk 8 above)
Comment (cosunae) [reply examples/python/tridiagonal_solve.py]: there were a couple of problems with a latest merge with master. Can you try again? 
You just need 
export PYTHONPATH=<wdir>/bundle/install/python

File: python/IIR/__init__.py (hunk 9 above)
Comment (cosunae) [reply python/IIR/__init__.py]: done

File: python/dawn/sir_printer.py (hunk 10 above)
Comment (cosunae) [reply python/dawn/sir_printer.py]: done

**Comment by cosunae** [comment]
launch jenkins

File: bundle/CMakeLists.txt [hunk 12]
```diff
@@ -64,10 +64,17 @@ yoda_add_target_clean_all(
...
+endif()
+
 include("thirdparty/DawnAddProtobuf")
+
+install(DIRECTORY ${PROTOBUF_PYTHON_INSTALL}/google DESTINATION ${CMAKE_INSTALL_PREFIX}/python)
+
 include("thirdparty/DawnAddBoost")
 
-set(dawn_cmake_args -DProtobuf_DIR=${Protobuf_DIR} -DBOOST_ROOT=${BOOST_ROOT})
+set(dawn_cmake_args -DProtobuf_DIR=${Protobuf_DIR} -DBOOST_ROOT=${BOOST_ROOT} -DDAWN_PYTHON=ON -DDAWN_EXAMPLES=ON)
```
Comment (twicki) [inline bundle/CMakeLists.txt]: I think that we should not build the examples by default. My suggestion is to have `make && make install` provide the default use-case with minimal building time. I know that the GT-team does this differently but their examples are basically codgen-tests. 

It is in my eyes debatable if we want to build DAWN_PYTHON by default (I would lean towards no). Happy to hear what people think. @Stagno do you have an opinion on this?

File: bundle/CMakeLists.txt [hunk 13]
```diff
@@ -97,3 +104,4 @@ install(
   DESTINATION bundle_conf
 )
 
+install (SCRIPT "${CMAKE_SOURCE_DIR}/PostInstall.cmake")
```
Comment (twicki) [inline bundle/CMakeLists.txt]: I think this should be a separate build step and not be done by default in `make` or `make install`

File: bundle/PostInstall.cmake [hunk 14]
```diff
@@ -0,0 +1,25 @@
+set(examples "copy_stencil;hori_diff;tridiagonal_solve")
```
Comment (twicki) [inline bundle/PostInstall.cmake]: when I run 
```
cmake ..
make install
```
this breaks. If using 
```
cmake .. -DDAWN_PYTHON=ON
make install
```
it works. we should probably check if it was set to on and do nothing otherwise?

File: examples/python/CMakeLists.txt [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+  message(WARNING 
+    "Python module of protobuf is not installed correctly! Expected protobuf's __init__.py at:\n"
+    "    ${protobuf_init_py}\n"
+    "Disabling python module of Dawn."
+  )
+else()
+  set(DAWN_INSTALL_PROTOBUF_MODULE "${protobuf_python}")
+
+  set(config_in "${CMAKE_SOURCE_DIR}/examples/python/config.py.in")
+  set(config_out "${CMAKE_BINARY_DIR}/examples/python/config.py")
```
Comment (twicki) [inline examples/python/CMakeLists.txt]: should the config be written to `${CMAKE_BINARY_DIR}/examples/python/config.py` and not to `${CMAKE_SOURCE_DIR}/examples/python/config.py`? I would expect to be able to run the examples built in the example directory and not in the install

File: examples/python/data/copy_stencil_ref.cpp [hunk 16]
```diff
@@ -0,0 +1,168 @@
+namespace cuda{
```
Comment (twicki) [inline examples/python/data/copy_stencil_ref.cpp]: is there a reason why we want the reference to be generated files rather than serialized IIR-files?

File: examples/python/hori_diff.py [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+This program creates the HIR corresponding to an horizontal diffusion stencil using the Python API of the HIR.
+The horizontal diffusion is a basic example that contains horizontal data dependencies that need to be resolved 
+by the compiler passes. 
+The code is meant as an example for high-level DSLs that could generate HIR from their own 
+internal IR. 
+The program contains two parts: 
+    1. construct the HIR of the example
+    2. pass the HIR to the dawn compiler in order to run all optimizer passes and code generation.
+       In this example the compiler is configured with the CUDA backend, therefore will code generate
+       an optimized CUDA implementation.
```
Comment (twicki) [inline examples/python/hori_diff.py]: should we specify the PYTHONPATH in this header?

File: examples/python/tridiagonal_solve.py (hunk 8 above)
Comment (twicki) [reply examples/python/tridiagonal_solve.py]: (repeated) should this be documented here?

File: src/dawn/IIR/CMakeLists.txt [hunk 18]
```diff
@@ -59,7 +62,7 @@ yoda_add_library(
           IntervalAlgorithms.h
           IIR.cpp
           IIR.h
-          IIR.proto
+	  proto/IIR/IIR.proto
```
Comment (twicki) [inline src/dawn/IIR/CMakeLists.txt]: indentation is broken (1 tab & 2 white space instead of the normal 10 white spaces we have normally)

File: src/dawn/IIR/CMakeLists.txt [hunk 19]
```diff
@@ -101,3 +105,40 @@ yoda_combine_libraries(
...
+  add_custom_target(DawnIIRPythonProto ALL
+    COMMAND ${CMAKE_COMMAND} -E copy ${iir_proto_python_files} ${CMAKE_BINARY_DIR}/python/dawn
+    DEPENDS ${iir_proto_python_files}
+  )
+  install(FILES ${iir_proto_python_files} DESTINATION ${DAWN_INSTALL_PYTHON_DIR}/IIR)
+  install(FILES ${CMAKE_SOURCE_DIR}/python/IIR/__init__.py DESTINATION ${DAWN_INSTALL_PYTHON_DIR}/IIR)
+endif(DAWN_PYTHON)
+
+# Genreate python proto files
+dawn_protobuf_generate(
```
Comment (twicki) [inline src/dawn/IIR/CMakeLists.txt]: should this have a guard that checks for a cmake-java-flag ?

File: src/dawn/SIR/CMakeLists.txt [hunk 20]
```diff
@@ -39,7 +41,7 @@ yoda_add_library(
           ASTVisitor.h
           SIR.cpp
           SIR.h
-          SIR.proto
+	  proto/SIR/SIR.proto
```
Comment (twicki) [inline src/dawn/SIR/CMakeLists.txt]: indentation again

File: src/dawn/SIR/CMakeLists.txt [hunk 21]
```diff
@@ -55,19 +57,45 @@ yoda_combine_libraries(
...
+  install(FILES ${sir_proto_python_files} DESTINATION ${DAWN_INSTALL_PYTHON_DIR}/SIR)
+  install(FILES ${CMAKE_SOURCE_DIR}/python/SIR/__init__.py DESTINATION ${DAWN_INSTALL_PYTHON_DIR}/SIR)
+endif(DAWN_PYTHON)
+
 # Genreate python proto files
 dawn_protobuf_generate(
-  OUT_FILES sir_proto_python_files
-  PROTOS ${CMAKE_CURRENT_SOURCE_DIR}/SIR.proto
-  LANGUAGE python
+  OUT_FILES sir_proto_java_files
```
Comment (twicki) [inline src/dawn/SIR/CMakeLists.txt]: java guard?

File: bundle/CMakeLists.txt (hunk 12 above)
Comment (cosunae) [reply bundle/CMakeLists.txt]: here we are using these examples as integration tests. If we dont build it and test it we are delivering a functionality that is always broken

File: examples/python/CMakeLists.txt (hunk 15 above)
Comment (cosunae) [reply examples/python/CMakeLists.txt]: I think in general is good practice to not touch the original source code during the build. 
The config also sets some directories to add to the python path, therefore it would be complicated to allow to build the module in multiple paths, so I would advise against

File: examples/python/data/copy_stencil_ref.cpp (hunk 16 above)
Comment (cosunae) [reply examples/python/data/copy_stencil_ref.cpp]: just for other users to be able to read this. But no strong reason

File: examples/python/hori_diff.py (hunk 17 above)
Comment (cosunae) [reply examples/python/hori_diff.py]: I think is a normal thing to expect (unless you go via pip) that for compiled and installed modules one has to set the PYTHONPATH to the directory of the module, isnt? 
So I would not repeat it in all examples... but we could put it in a README or general documentation of how to use python

File: src/dawn/IIR/CMakeLists.txt (hunk 18 above)
Comment (cosunae) [reply src/dawn/IIR/CMakeLists.txt]: fixed

File: src/dawn/SIR/CMakeLists.txt (hunk 20 above)
Comment (cosunae) [reply src/dawn/SIR/CMakeLists.txt]: fixed

File: src/dawn/IIR/CMakeLists.txt (hunk 19 above)
Comment (cosunae) [reply src/dawn/IIR/CMakeLists.txt]: ok, done

File: bundle/CMakeLists.txt (hunk 12 above)
Comment (cosunae) [reply bundle/CMakeLists.txt]: or we could set the default in the options to OFF, and in the script of jenkins we activate it. 
Let me give it a try if you are happy with this solution

**Comment by cosunae** [comment]
@twicki can you have a look again?

**Comment by cosunae** [comment]
notice there was a problem with the python install of protobuf, so that is why it was not working for you probably. I fixed that in yoda

**Comment by cosunae** [comment]
launch jenkins

**Comment by cosunae** [comment]
launch jenkins

**Comment by cosunae** [comment]
launch jenkins

File: bundle/CMakeLists.txt (hunk 12 above)
Comment (twicki) [reply bundle/CMakeLists.txt]: That sounds like the best approach to me. Building is still on the "minimal path" and testing is more extended

File: examples/python/hori_diff.py (hunk 17 above)
Comment (twicki) [reply examples/python/hori_diff.py]: I would probably just write a couple of lines somewhere. I agree repreating it everywhere might be suboptimal

File: bundle/CMakeLists.txt [hunk 22]
```diff
@@ -64,17 +64,22 @@ yoda_add_target_clean_all(
...
+option(DAWN_BUNDLE_PYTHON "Build and install the Python module interface to HIR" OFF)
+option(DAWN_BUNDLE_JAVA "Build and install the java interface to HIR" OFF)
+
+if( DAWN_BUNDLE_PYTHON )
+  if(NOT DEFINED PYTHON_EXECUTABLE)
+    find_package(PythonInterp 3.4 REQUIRED)
+  endif()
+endif()
+
+set(dawn_cmake_args -DProtobuf_DIR=${Protobuf_DIR} -DBOOST_ROOT=${BOOST_ROOT} -DDAWN_PYTHON=${DAWN_BUNDLE_PYTHON} -DDAWN_EXAMPLES=ON -DDAWN_JAVA=${DAWN_BUNDLE_JAVA})
```
Comment (twicki) [inline bundle/CMakeLists.txt]: Is the plan to extend the dawn examples so there is no hard requirement between examples and python? I think we can leave it like that. Just curious on what you think where this should go?

File: scripts/jenkins/build.sh [hunk 23]
```diff
@@ -18,7 +18,7 @@ if [ -z ${PROTOBUFDIR+x} ]; then
  echo "PROTOBUFDIF needs to be set in the machine env"
 fi
 
-cmake -DCMAKE_BUILD_TYPE=${build_type} -DBOOST_ROOT=${BOOST_DIR}  \
+cmake -DDAWN_BUNDLE_PYTHON=ON -DDAWN_BUNDLE_JAVA=ON -DCMAKE_BUILD_TYPE=${build_type} -DBOOST_ROOT=${BOOST_DIR}  \
```
Comment (twicki) [inline scripts/jenkins/build.sh]: I think this is a good choice

File: examples/python/CMakeLists.txt (hunk 15 above)
Comment (twicki) [reply examples/python/CMakeLists.txt]: ok, I agree

File: examples/python/data/copy_stencil_ref.cpp (hunk 16 above)
Comment (twicki) [reply examples/python/data/copy_stencil_ref.cpp]: sounds like a good reason to me

File: bundle/CMakeLists.txt (hunk 13 above)
Comment (twicki) [reply bundle/CMakeLists.txt]: can we move this to a test similar to https://github.com/MeteoSwiss-APN/gtclang/blob/master/test/utils/gtclang-tester/CMakeLists.txt

Comment (twicki) [reply bundle/CMakeLists.txt]: I was forgetting the problem of having to do it at install-time....
```suggestion
if( DAWN_BUNDLE_PYTHON )
  install (SCRIPT "${CMAKE_SOURCE_DIR}/PostInstall.cmake")
endif()
```

**Comment by twicki** [comment]
adding the guard around the cmake (as suggested) and then we can merge
